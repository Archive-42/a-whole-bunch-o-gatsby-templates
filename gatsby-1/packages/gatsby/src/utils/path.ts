import path from "path"
import { joinPath, createContentDigest } from "gatsby-core-utils"

export const withBasePath = (basePath: string) => (
  ...paths: Array<string>
): string => joinPath(basePath, ...paths)

export const withTrailingSlash = (basePath: string): string => `${basePath}/`

const posixJoinWithLeadingSlash = (paths: Array<string>): string =>
  path.posix.join(
    ...paths.map((segment, index) =>
      segment === `` && index === 0 ? `/` : segment
    )
  )

export const getCommonDir = (path1: string, path2: string): string => {
  const path1Segments = path1.split(/[/\\]/)
  const path2Segments = path2.split(/[/\\]/)

  for (let i = 0; i < path1Segments.length; i++) {
    if (i >= path2Segments.length) {
      return posixJoinWithLeadingSlash(path2Segments)
    } else if (
      path1Segments[i].toLowerCase() !== path2Segments[i].toLowerCase()
    ) {
      const joined = path1Segments.slice(0, i)
      return posixJoinWithLeadingSlash(joined)
    }
  }

  return posixJoinWithLeadingSlash(path1Segments)
}

// MacOS (APFS) and Windows (NTFS) filename length limit = 255 chars, Others = 255 bytes
const MAX_PATH_SEGMENT_CHARS = 255
const MAX_PATH_SEGMENT_BYTES = 255
const SLICING_INDEX = 50
const pathSegmentRe = /[^/]+/g

const isMacOs = process.platform === `darwin`
const isWindows = process.platform === `win32`

const isNameTooLong = (segment: string): boolean =>
  isMacOs || isWindows
    ? segment.length > MAX_PATH_SEGMENT_CHARS // MacOS (APFS) and Windows (NTFS) filename length limit (255 chars)
    : Buffer.from(segment).length > MAX_PATH_SEGMENT_BYTES // Other (255 bytes)

export const tooLongSegmentsInPath = (path: string): Array<string> => {
  const invalidFilenames: Array<string> = []
  for (const segment of path.split(`/`)) {
    if (isNameTooLong(segment)) {
      invalidFilenames.push(segment)
    }
  }
  return invalidFilenames
}

export const truncatePath = (path: string): string =>
  path.replace(pathSegmentRe, match => {
    if (isNameTooLong(match)) {
      return (
        match.slice(0, SLICING_INDEX) +
        createContentDigest(match.slice(SLICING_INDEX))
      )
    }
    return match
  })
