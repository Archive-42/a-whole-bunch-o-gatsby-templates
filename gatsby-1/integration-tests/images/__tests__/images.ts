import { readFile, writeFile, ensureDir } from "fs-extra"
import path from "path"
import { JSDOM } from "jsdom"
import sharp from "sharp"
import pixelmatch from "pixelmatch"
import { PNG } from "pngjs"

type ImgData = { metadata: sharp.Metadata; data: Buffer }

/**
 * Loads an image from disk and converts it to a raw pixel buffer
 */
async function toImageBuffer(imagePath: string): Promise<ImgData> {
  const filename = path.join(__dirname, "..", "public", imagePath)
  const pipeline = sharp(filename)
  const metadata = await pipeline.metadata()
  const data = await pipeline.ensureAlpha().raw().toBuffer()
  return { metadata, data }
}

/**
 *  Extracts image path/size pairs from a srcset
 */
function parseSrcSet(srcset: string): Array<{ src: string; size: number }> {
  return srcset.split(",").map(item => {
    const [src, size] = item.split(" ")
    return { src: src.trim(), size: parseInt(size) }
  })
}

type Source = {
  format: string
  srcset: Array<{ src: string; size: number }>
}

/**
 * Loads all image srcsets below the given testId
 */
function getImages(testId: string, dom: HTMLDivElement) {
  const node = dom.querySelector(`[data-test-id="${testId}"]`)

  const sources = node?.querySelectorAll("source")

  const img = node?.querySelector<HTMLImageElement>("[data-main-image]")

  const srcsets: Array<Source> = []
  const imgsrcset = img?.dataset.srcset

  if (imgsrcset) {
    srcsets.push({ format: "auto", srcset: parseSrcSet(imgsrcset) })
  }

  sources?.forEach(source => {
    const format = source.getAttribute("type")
    const srcset = source.dataset.srcset
    if (format && srcset) {
      srcsets.push({ format, srcset: parseSrcSet(srcset) })
    }
  })
  return srcsets
}

/**
 * Performs a pixel-level comparison on two images. Fails if there is more than 1% difference.
 * This is enough to catch common errors such as issues with cropping, without tripping on
 * differences in image quality, artifacts etc.
 */
async function imageDiff(source: ImgData, target: string): Promise<boolean> {
  const targetData = await toImageBuffer(target)
  const { metadata, data } = targetData

  if (
    metadata.width !== source.metadata.width ||
    metadata.height !== source.metadata.height
  ) {
    console.warn("Image sizes don't match")
    return false
  }

  const diffImg = new PNG(metadata)

  const diff = pixelmatch(
    data,
    source.data,
    diffImg.data,
    metadata.width,
    metadata.height,
    { threshold: 0.2 }
  )

  const pass = diff < data.length / 4 / 100 //1% change

  if (!pass) {
    const name = `${expect.getState().currentTestName} ${path.basename(
      target
    )} ${source.metadata.width}x${source.metadata.height}`.replace(/\W+/g, "_")

    await ensureDir("__diff_output__")
    await writeFile(`__diff_output__/${name}.png`, PNG.sync.write(diffImg))
  }

  return pass
}

/**
 * Loads an HTML file and returns the root gatsby DOM element
 */
async function loadHTML(pathName): Promise<HTMLDivElement> {
  const html = await readFile(
    path.resolve(__dirname, "..", "public", pathName),
    "utf-8"
  )
  return new JSDOM(html).window.document.getElementById("___gatsby")
}

/**
 * Given a pair of srcsets, returns an array of mismatching images
 */

async function compareSrcSets(
  main: Source,
  other: Source
): Promise<Array<string>> {
  const results = await Promise.all(
    main.srcset.map(
      async ({ src, size }, index): Promise<string | null> => {
        const title = `${path.basename(src)} does not match ${path.basename(
          other.srcset[index].src
        )} at size ${size}`
        const source = await toImageBuffer(src)
        const result = await imageDiff(source, other.srcset[index].src)
        if (!result) {
          return title
        }
        return null
      }
    )
  )
  return results.filter(Boolean) as Array<string>
}

declare global {
  namespace jest {
    interface VisualMatch extends JestMatchers<Source> {
      toVisuallyMatchSrcSet: (
        comparison: Source
      ) => Promise<jest.CustomMatcherResult>
    }
  }
}

expect.extend({
  toVisuallyMatchSrcSet: async (received: Source, comparison: Source) => {
    const results = await compareSrcSets(comparison, received)
    if (!results.length) {
      return {
        pass: true,
        message: () =>
          `expected ${received.format} srcset to not match original`,
      }
    }
    return {
      pass: false,
      message: () =>
        `expected ${
          received.format
        } srcset to match original:\n\n${results.join(
          "\n"
        )}\n\nSee "__diff_output__" for comparison`,
    }
  },
})

let div: HTMLDivElement

describe("image generation", () => {
  beforeAll(async () => {
    div = await loadHTML("index.html")
  })

  it("generates the correct images for test image", async () => {
    return Promise.all(
      [
        "constrained-square",
        "forcePNG",
        "aspectRatio",
        "fullWidth-aspectRatio",
        "fixed-aspectRatio",
        "fixed",
        "fixed-square",
        "constrained",
        "no-jpg",
        "fullwidth",
        "fullwidth-bp",
      ].map(async testId => {
        const images = getImages(testId, div)

        expect(images).toMatchSnapshot(testId)

        const [fallback, ...imgs] = images

        await Promise.all(
          imgs.map(img => {
            return (expect(img) as jest.VisualMatch).toVisuallyMatchSrcSet(
              fallback
            )
          })
        )
      })
    )
  }, 60000)
})
