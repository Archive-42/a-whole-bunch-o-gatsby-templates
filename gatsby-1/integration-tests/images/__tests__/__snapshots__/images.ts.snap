// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image generation generates the correct images for test image: aspectRatio 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 125,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/e260b/sighisoara.jpg",
      },
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/49f40/sighisoara.jpg",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/8b0b1/sighisoara.jpg",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/6bc8e/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 125,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/775a0/sighisoara.avif",
      },
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/0871e/sighisoara.avif",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/88400/sighisoara.avif",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/fd21d/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 125,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/e40ff/sighisoara.webp",
      },
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/1828c/sighisoara.webp",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/9f72e/sighisoara.webp",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/aa61a/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: constrained 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/99f43/sighisoara.jpg",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/26cfb/sighisoara.jpg",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/1b035/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/ea46c/sighisoara.avif",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/569a0/sighisoara.avif",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/ea316/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/48571/sighisoara.webp",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/921bf/sighisoara.webp",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/b7ef5/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: constrained-square 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 125,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/8bb35/sighisoara.jpg",
      },
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/f505e/sighisoara.jpg",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/be5ed/sighisoara.jpg",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/5a7c3/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 125,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/3ebc4/sighisoara.avif",
      },
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/a9ff4/sighisoara.avif",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/a61b1/sighisoara.avif",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/9ef3c/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 125,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/d66e1/sighisoara.webp",
      },
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/e7160/sighisoara.webp",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/5f169/sighisoara.webp",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/3cd29/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: fixed 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/1b035/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/ea316/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/b7ef5/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: fixed-aspectRatio 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/8b0b1/sighisoara.jpg",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/6bc8e/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/88400/sighisoara.avif",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/fd21d/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/9f72e/sighisoara.webp",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/aa61a/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: fixed-square 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 300,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/0fdf4/sighisoara.jpg",
      },
      Object {
        "size": 600,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/a89ca/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 300,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/aa5b9/sighisoara.avif",
      },
      Object {
        "size": 600,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/0c8d0/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 300,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/078c3/sighisoara.webp",
      },
      Object {
        "size": 600,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/6d09e/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: forcePNG 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 125,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/8bb35/sighisoara.jpg",
      },
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/f505e/sighisoara.jpg",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/be5ed/sighisoara.jpg",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/5a7c3/sighisoara.jpg",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: fullWidth-aspectRatio 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/5d30b/sighisoara.jpg",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/29c4b/sighisoara.jpg",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/6bfa7/sighisoara.jpg",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/0772e/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/c92f7/sighisoara.avif",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/bd57f/sighisoara.avif",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/d9624/sighisoara.avif",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/fbe86/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/316e8/sighisoara.webp",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/8386b/sighisoara.webp",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/46618/sighisoara.webp",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/15204/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: fullwidth 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/7284f/sighisoara.jpg",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/29ba9/sighisoara.jpg",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/c8896/sighisoara.jpg",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/5267c/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/b0356/sighisoara.avif",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/48f75/sighisoara.avif",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/a5094/sighisoara.avif",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/f46cf/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/57584/sighisoara.webp",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/984df/sighisoara.webp",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/4a276/sighisoara.webp",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/9c00f/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: fullwidth-bp 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/7284f/sighisoara.jpg",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/29ba9/sighisoara.jpg",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/c8896/sighisoara.jpg",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/5267c/sighisoara.jpg",
      },
    ],
  },
  Object {
    "format": "image/avif",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/b0356/sighisoara.avif",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/48f75/sighisoara.avif",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/a5094/sighisoara.avif",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/f46cf/sighisoara.avif",
      },
    ],
  },
  Object {
    "format": "image/webp",
    "srcset": Array [
      Object {
        "size": 750,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/57584/sighisoara.webp",
      },
      Object {
        "size": 1080,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/984df/sighisoara.webp",
      },
      Object {
        "size": 1366,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/4a276/sighisoara.webp",
      },
      Object {
        "size": 1920,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/9c00f/sighisoara.webp",
      },
    ],
  },
]
`;

exports[`image generation generates the correct images for test image: no-jpg 1`] = `
Array [
  Object {
    "format": "auto",
    "srcset": Array [
      Object {
        "size": 250,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/48571/sighisoara.webp",
      },
      Object {
        "size": 500,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/921bf/sighisoara.webp",
      },
      Object {
        "size": 1000,
        "src": "/static/3dc14781080130c49ec2f3d768d5d244/b7ef5/sighisoara.webp",
      },
    ],
  },
]
`;
